src/technochat/http/ochat.go

```diff
@@ -0,0 +1,200 @@
+	Username string
+}
+
+type TypeID int
+
+const (
+	WSMsgTypeService TypeID = iota
+	WSMsgTypeMessage
+)
+
+type Event struct {
+	EventID   EventID     `json:"event_id"`
+	EventData interface{} `json:"event_data"`
+}
+type EventID int
+
+const (
+	EventConnInitOk EventID = iota
+	EventConnInitNoSuchChat
+	EventConnInitMaxUsrReached
+)
+
+type WSMessage struct {
+	Type     TypeID      `json:"type"`
+	Data     interface{} `json:"data"`
+	Username string      `json:"username"`
+}
+
+type OchatInitRequest struct {
+	MaxUsers int `json:"max_users"`
+}
+type OchatInitResponse struct {
+	Id string `json:"id"`
+}
+
+func (s *Server) ochatInit(r *http.Request) (int, interface{}, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return http.StatusBadRequest, nil, err
+	}
+
+	req := OchatInitRequest{MaxUsers: DefaultMaxPPL}
+	if len(body) != 0 {
+		err = json.Unmarshal(body, &req)
+		if err != nil {
+			return http.StatusBadRequest, nil, err
+		}
+	}
+
+	//generate chat id
+	chatID, err := uuid.NewRandom()
+	if err != nil {
+		return http.StatusBadRequest, nil, err
+	}
+	chatIDStr := chatID.String()
+
+	//init chat
+	chats[chatIDStr] = &Chat{
+		corresps:  make(map[*User]bool),
+		broadcast: make(chan WSMessage),
+		usedNames: make(map[int]bool),
+		pplToJoin: req.MaxUsers,
+	}
+
+	//do broadcasts
+	go func() {
+		for {
+			select {
+			case <-time.After(360 * time.Second):
+				for usr := range chats[chatIDStr].corresps { //TODO: refactor
+					usr.WS.Close()
+					delete(chats[chatIDStr].corresps, usr)
+				}
+				delete(chats, chatIDStr)
+				return
+			case msg := <-chats[chatIDStr].broadcast:
+				for usr := range chats[chatIDStr].corresps {
+					err := usr.WS.WriteJSON(msg)
+					if err != nil {
+						log.Printf("error broadcasting: %v", err)
+						usr.WS.Close()
+						delete(chats[chatIDStr].corresps, usr)
+					}
+				}
+			}
+		}
+	}()
+
+	resp := map[string]string{
+		"id": chatIDStr,
+	}
+	return http.StatusOK, resp, nil
+}
+
+func (s *Server) ochatConnect(w http.ResponseWriter, r *http.Request) {
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		log.Printf("error upgrading chat connection: %v", err)
+	}
+	defer ws.Close()
+
+	err = r.ParseForm()
+	if err != nil {
+		log.Printf("error parsing form :%v", err)
+	}
+	chatIDStr := r.FormValue("id")
+
+	//check chat existance
+	chat, ok := chats[chatIDStr]
+	if !ok {
+		log.Printf("chat %s not exists", chatIDStr)
+		sendEvent(ws, EventConnInitNoSuchChat, nil)
+		return
+	} else if chat.pplToJoin <= 0 {
+		log.Printf("chat %s max user reached", chatIDStr)
+		sendEvent(ws, EventConnInitNoSuchChat, nil)
+		return
+	}
+
+	usr := chats[chatIDStr].AddUser(ws)
+	sendEvent(usr.WS, EventConnInitOk, usr.Username)
+
+	sendServerNotify(chats[chatIDStr].broadcast, "user "+usr.Username+" joined")
+	chats[chatIDStr].pplToJoin--
+	for {
+		msg := WSMessage{
+			Username: usr.Username,
+		}
+		err := usr.WS.ReadJSON(&msg)
+		if err != nil {
+			log.Printf("error: %v", err)
```

Comment:
`error: could not read message from user %s: %v`

Reply:
done